Pems Data (#29)
deephaven/examples: pems/README.md

> # PEMS Example data

Reviewer: ```suggestion
# PeMS example data
```

This folder contains data files concerning [Caltrans Performance Measurement System (PeMS)](https://pems.dot.ca.gov/) traffic flow.

> These partitioned parquet files are provided by [Clark Fitzgerald](https://anson.ucdavis.edu/~clarkf/) and show the ability to read such structured files in the Deephaven Community Edition.

Reviewer: ```suggestion
These partitioned Parquet files are provided by [Clark Fitzgerald](https://anson.ucdavis.edu/~clarkf/) and show the ability to read such structured files in Deephaven Community Core.
```

> The data represent aggregate traffic flow collected every 30 second from raw sensors on the three lanes of traffic from January to October 2016 for a location of the I80 corridor near Davis, CA.

Reviewer: ```suggestion
The data represent aggregate traffic flow collected every 30 seconds from raw sensors on the three lanes of traffic, from January to October 2016, for a location of the I-80 corridor near Davis, CA.

```

> PeMS computes performance measures from sensor measurements of vehicle count (flow), occupancy, and speed automatically collected on the freeway.

Reviewer: ```suggestion
PeMS computes performance measures from sensor measurements of vehicle count (flow), occupancy, and speed automatically collected on the freeway.
```

> Occupancy is a measure of how much of the roadway is covered by vehicles.  A related quantity is the freeway density, measured in vehicles per mile.

Reviewer: ```suggestion
- Occupancy is a measure of how much of the roadway is covered by vehicles.  
- A related quantity is the freeway density, measured in vehicles per mile.
```

> In this location there are only three lanes so only three sensors of information are provided. Lanes 5-8 are present in the _meta_common to facilitate import other locations into the dataset.

Reviewer: ```suggestion
In this location there are only three lanes so only three sensors of information are provided. Lanes 5-8 are present in the `_meta_common` to facilitate importing other locations into the dataset.
```

## Table of contents

> - `parquet`: directory containing partitioned parquet data

Reviewer: ```suggestion
- `parquet`: directory containing partitioned Parquet data
```

## Fields

  - **timeperiod (String):**  Date and time value for the sample period.

>   - **flow1 (int):**  The number of vehicles observed in lane 1 for sample period.

Reviewer: ```suggestion
  - **flow1 (int):**  The number of vehicles observed in lane 1 for the sample period.
```

  - **occupancy1 (double):** The fraction of each sample period during which the detector in lane 1 output is “1”.
  - **speed1 (double):**  The average speed of vehicles in lane 1 during the sample period.

>   - **flow2 (int):**  The number of vehicles observed in lane 2 for sample period.

Reviewer: ```suggestion
  - **flow2 (int):**  The number of vehicles observed in lane 2 for the sample period.
```

  - **occupancy2 (double):** The fraction of each sample period during which the detector in lane 2 output is “1”.
  - **speed2 (double):**  The average speed of vehicles in lane 2 during the sample period.

>   - **flow3 (int):**  The number of vehicles observed in lane 3 for sample period.

Reviewer: ```suggestion
  - **flow3 (int):**  The number of vehicles observed in lane 3 for the sample period.
```

  - **occupancy3 (double):** The fraction of each sample period during which the detector in lane 3 output is “1”.
  - **speed3 (double):**  The average speed of vehicles in lane 3 during the sample period.

>   - **flow4 (int):**  Null for this location. The number of vehicles observed in lane 4 for sample period.

Reviewer: ```suggestion
  - **flow4 (int):**  Null for this location. The number of vehicles observed in lane 4 for the sample period.
```

  - **occupancy4 (double):** Null for this location. The fraction of each sample period during which the detector in lane 4 output is “1”.
  - **speed4 (double):**  Null for this location. The average speed of vehicles in lane 4 during the sample period.

>   - **flow5 (int):**  Null for this location. The number of vehicles observed in lane 5 for sample period.

Reviewer: ```suggestion
  - **flow5 (int):**  Null for this location. The number of vehicles observed in lane 5 for the sample period.
```

  - **occupancy5 (double):** Null for this location. The fraction of each sample period during which the detector in lane 5 output is “1”.
  - **speed5 (double):**  Null for this location. The average speed of vehicles in lane 5 during the sample period.

>   - **flow6 (int):**  Null for this location. The number of vehicles observed in lane 6 for sample period.

Reviewer: ```suggestion
  - **flow6 (int):**  Null for this location. The number of vehicles observed in lane 6 for the sample period.
```

  - **occupancy6 (double):** Null for this location. The fraction of each sample period during which the detector in lane 6 output is “1”.
  - **speed6 (double):**  Null for this location. The average speed of vehicles in lane 6 during the sample period.

>   - **flow7 (int):**  Null for this location. The number of vehicles observed in lane 7 for sample period.

Reviewer: ```suggestion
  - **flow7 (int):**  Null for this location. The number of vehicles observed in lane 7 for the sample period.
```

  - **occupancy7 (double):** Null for this location. The fraction of each sample period during which the detector in lane 7 output is “1”.
  - **speed7 (double):**  Null for this location. The average speed of vehicles in lane 7 during the sample period.

>   - **flow8 (int):**  Null for this location. The number of vehicles observed in lane 8 for sample period.

Reviewer: ```suggestion
  - **flow8 (int):**  Null for this location. The number of vehicles observed in lane 8 for the sample period.
```

# PeMS example data

This folder contains data files concerning [Caltrans Performance Measurement System (PeMS)](https://pems.dot.ca.gov/) traffic flow.
These partitioned Parquet files are provided by [Clark Fitzgerald](https://anson.ucdavis.edu/~clarkf/) and show the ability to read such structured files in Deephaven Community Core.

> The data represent aggregate traffic flow collected every 30 seconds from raw sensors on the three lanes of traffic, from January to October 2016, for a location of the I-80 corridor near Davis, CA.

Reviewer: ```suggestion
The data represents aggregate traffic flow collected every 30 seconds from raw sensors on the three lanes of traffic, from January to October 2016, for a location of the I-80 corridor near Davis, CA.
```

  - **flow6 (int):**  Null for this location. The number of vehicles observed in lane 6 for the sample period.
  - **occupancy6 (double):** Null for this location. The fraction of each sample period during which the detector in lane 6 output is “1”.
  - **speed6 (double):**  Null for this location. The average speed of vehicles in lane 6 during the sample period.
  - **flow7 (int):**  Null for this location. The number of vehicles observed in lane 7 for the sample period.
  - **occupancy7 (double):** Null for this location. The fraction of each sample period during which the detector in lane 7 output is “1”.
  - **speed7 (double):**  Null for this location. The average speed of vehicles in lane 7 during the sample period.
  - **flow8 (int):**  Null for this location. The number of vehicles observed in lane 8 for the sample period.
  - **occupancy8 (double):** Null for this location. The fraction of each sample period during which the detector in lane 8 output is “1”.
  - **speed8 (double):**  Null for this location. The average speed of vehicles in lane 8 during the sample period.

# Source and License

> This data was built from data sets publicly available on [Caltrans Performance Measurement System (PeMS)](https://pems.dot.ca.gov/). It is provided here for demonstrative use without any warranty as to the accuracy, reliability, or completeness of the data.

Reviewer 2: A few questions:
1) How big is this data set?  I see 1425 files.
2) Is `./pems/parquet/pems/_SUCCESS` needed  It is an empty file?  If it is needed, what does it do?
3) Are `./pems/parquet/pems/_common_metadata` and `./pems/parquet/pems/_metadata` Deephaven created files?  Do they come from somewhere else?
4) For the files in 2 & 3 above, do we need to document anything about them in the readme?  I clearly don't know what they do, so I'm guessing others won't either.

Author: 1) 43.2 MB for 1,430 items including _* and README
2) did not need _SUCCESS
3) These files are useful for reading, one of the reasons for switching to this data set is to talk about them in the docs and users can see the files, while the other data set did not have them.
4) added.

# PeMS example data

This folder contains data files concerning [Caltrans Performance Measurement System (PeMS)](https://pems.dot.ca.gov/) traffic flow.
These partitioned Parquet files are provided by [Clark Fitzgerald](https://anson.ucdavis.edu/~clarkf/) and show the ability to read such structured files in Deephaven Community Core.

The data represents aggregate traffic flow collected every 30 seconds from raw sensors on the three lanes of traffic, from January to October 2016, for a location of the I-80 corridor near Davis, CA.

PeMS computes performance measures from sensor measurements of vehicle count (flow), occupancy, and speed automatically collected on the freeway.
- Occupancy is a measure of how much of the roadway is covered by vehicles.  
- A related quantity is the freeway density, measured in vehicles per mile.
In this location there are only three lanes so only three sensors of information are provided. Lanes 5-8 are present in the `_meta_common` to facilitate importing other locations into the dataset.

## Table of contents

> - `parquet`: directory containing partitioned Parquet data

Reviewer: ```suggestion
- `parquet`: Directory containing partitioned Parquet data.
```

> - `_common_metadata`: file contains the schema for the whole dataset, including those hierarchical columns and those not in this specific section of the data.

Reviewer: ```suggestion
- `_common_metadata`: File contains the schema for the whole dataset, including those hierarchical columns and those not in this specific section of the data.
```

> - `_common_metadata`: File contains information to ease reading the whole dataset without reading column names from individual files. Including column headers from whole dataset and those just this specific section of the data.

Reviewer: ```suggestion
- `_common_metadata`: File containing information to ease reading the whole dataset without reading column names from individual files, including column headers from the whole dataset and just those from this specific section of the data.
```

> - `_metadata`: File containing (1) complete relative pathnames to individual data files and (2) column statistics, such as min, max, etc., for the individual data files.

Reviewer: ```suggestion
- `_metadata`: File containing (1) complete relative pathnames to individual data files, and (2) column statistics, such as min, max, etc., for the individual data files.
```

> - `_metadata`: File contains column statistics such as min, max, etc. for the individual files, along with complete relative path names.

Reviewer: ```suggestion
- `_metadata`: File containing column statistics such as min, max, etc., for the individual files, along with complete relative path names.
```

> - `_metadata`: file contains the schema used for part files (omitting the hierarchical columns) and also includes per-file column stats (min, max, etc) for all the files, with their complete relative path names.

Reviewer: ```suggestion
- `_metadata`: File contains the schema used for part files (omitting the hierarchical columns) and per-file column stats (min, max, etc) for all the files, with their complete relative path names.
```

## Table of contents

- `parquet`: Directory containing partitioned Parquet data.
- `_common_metadata`: File contains the schema for the whole dataset, including those hierarchical columns and those not in this specific section of the data.

> - `_metadata`: File contains the schema used for part files (omitting the hierarchical columns) and per-file column stats (min, max, etc.,) for all the files, with their complete relative path names.

Reviewer 2: These descriptions contain lingo that someone that knows the details of what is going on would understand, but I'm left trying to guess.  Can we rephrase these to hit readers that aren't as educated on what is going on?

## Table of contents

- `parquet`: Directory containing partitioned Parquet data.

> - `_common_metadata`: File containing information to ease reading the whole dataset without reading column names from individual files, including column headers from the whole dataset and just those from this specific section of the data.

Reviewer 2: ```suggestion
- `_common_metadata`: File containing schema information needed to load the whole dataset faster.  
```

> - `_metadata`: File containing column statistics such as min, max, etc., for the individual files, along with complete relative path names.

Reviewer 2: ```suggestion
- `_metadata`: File containing (1) complete relative pathnames to individual data files and (2) column statistics, such as min, max, etc., for the individual data files.
```
